Implement script.getRealms for main window contexts

sergedeh opened the pull request:
#### 44db1f726520a1482bcbd375d5bdb588c4c3f7b5
<pre>
Implement script.getRealms for main window contexts
<a href="https://bugs.webkit.org/show_bug.cgi?id=303926">https://bugs.webkit.org/show_bug.cgi?id=303926</a>

Reviewed by BJ Burg.

This change implements the WebDriver BiDi script.getRealms command for main window browsing contexts, enabling tests to query JavaScript realms in the page.

Added RealmInfo, RealmType types to BidiScript.json protocol, implemented getRealms() in BidiScriptAgent for main window contexts with realm ID generation and navigation detection. Added basic realm ID support to evaluate() and callFunction().

Note: Realm ID consistency between evaluate() and getRealms() is not fully implemented - this will be addressed in a follow-up change.

* Source/WebKit/UIProcess/Automation/BidiScriptAgent.h:
* Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
(WebKit::BidiScriptAgent::callFunction):
(WebKit::BidiScriptAgent::evaluate):
(WebKit::BidiScriptAgent::getRealms):
(WebKit::BidiScriptAgent::createRealmInfoForFrame):
(WebKit::BidiScriptAgent::generateRealmIdForFrame):
(WebKit::BidiScriptAgent::generateRealmIdForBrowsingContext):
(WebKit::BidiScriptAgent::processRealmsForPagesAsync):
(WebKit::BidiScriptAgent::contextHandleForFrame):
(WebKit::BidiScriptAgent::collectExecutionReadyFrameRealms):
* Source/WebKit/UIProcess/Automation/protocol/BidiScript.json:
* WebDriverTests/TestExpectations.json:

Canonical link: <a href="https://commits.webkit.org/304800@main">https://commits.webkit.org/304800@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/402157a9a0b5e0784caa3d80233e322842c2710f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136509 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8866 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47789 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144222 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9554 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8710 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104390 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139454 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6977 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122324 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85225 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6619 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4299 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4815 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115937 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146971 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8548 "Built successfully") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41087 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 4 flakes 1 failures; Uploaded test results; Running re-run-layout-tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112730 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8565 "Built successfully") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7186 "Build is in progress. Recent messages:OS: Tahoe (26.2), Xcode: 26.2; Running apply-patch; Checked out pull request; Skipped layout-tests; 11 flakes 4 failures; Uploaded test results; Running re-run-layout-tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113074 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28724 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6556 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118621 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62545 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8596 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36672 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8315 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72171 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8536 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8388 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,311 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+        for (auto& realmInfo : accumulated) {
+            if (!realmInfo)
+                continue;
+            if (optionalType && *optionalType != Inspector::Protocol::BidiScript::RealmType::Window)
```
`std::optional<T>::operator!=` does this for you actually
```suggestion
            if (optionalType != Inspector::Protocol::BidiScript::RealmType::Window)
```

sergedeh replied:
I was following that part of the BiDi spec [here](https://w3c.github.io/webdriver-bidi/#command-script-getRealms) :

> If command parameters contains type and realm info["type"] is not equal to command parameters["type"] then continue.

In the case we have no type given in input (Type = nullopt), if we use the operator  `std::optional<T>::operator!=`  the line 366 will be true ( nullopt != Window return true) and no realm will be output.
The current implementation follow the path such that if no type is given as input, we should output all the realms (no filtering). I think it follows the BiDi two parts condition to the letter.

aprotyas replied:
ah okay, I misunderstood that intent.

the-chenergy commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,311 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+        // Don't treat this as a state change that increments the counter.
+        bool transitioningToFromBlank = (cachedURL != currentURL)
+            && ((cachedURL == "about:blank"_s && currentURL != "about:blank"_s)
+            || (cachedURL != "about:blank"_s && currentURL == "about:blank"_s));
```
```suggestion
        bool transitioningToOrFromBlank = (cachedURL == "about:blank"_s) != (currentURL == "about:blank"_s);
```

the-chenergy commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,311 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+        if (transitioningToFromBlank) {
+            m_frameRealmCache.remove(frameInfo.frameID);
+            m_frameRealmCounters.remove(frameInfo.frameID);
+            it = m_frameRealmCache.end();
```
Why this assignment, did you want to ensure the `it` wasn't some invalid stale value after the removal? If so, perhaps making `it` more local when declaring it would be cleaner:
```
    if (auto it = m_frameRealmCache.find(frameInfo.frameID); it != m_frameRtealmCache.end()) {
```

the-chenergy commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,311 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+    auto it = m_frameRealmCache.find(frameInfo.frameID);
+    if (it != m_frameRealmCache.end()) {
+        const String& cachedStateSignature = it->value.first;
+        const String& cachedRealmId = it->value.second;
```
```suggestion
        const auto& [cachedStateSignature, cachedRealmId] = it->value;
```

AnujPanta1 commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,311 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+
+    // Process the first page and recursively handle the rest
+    Ref<WebPageProxy> currentPage = pagesToProcess[0];
+    pagesToProcess.removeAt(0);
```
`Vector::removeAt(0)` is `O(n)` because it shifts all elements. For a recursive function processing N pages, this makes the overall complexity `O(n²)`

AnujPanta1 replied:
passing an index might work, or a reverse vector using `takeLast()`

the-chenergy replied:
Or use `WTF::Deque` instead of `WTF::Vector`

the-chenergy commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,311 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+        } else {
+            if (auto realmInfo = createRealmInfoForFrame(frameTree.info))
+                realms.append(realmInfo);
+        }
```
```suggestion
        bool shouldInclude = !contextHandleFilter || contextHandleForFrame(frameTree.info) == *contextHandleFilter;
        if (shouldInclude) {
            if (auto realmInfo = createRealmInfoForFrame(frameTree.info))
                realms.append(realmInfo);
        }
```

AnujPanta1 commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,311 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+    // We include documentID to detect reloads (same URL, different document).
+    // The documentID uniquely identifies a document instance, so navigating to the same URL
+    // (e.g., about:blank → about:blank) will have a different documentID and create a new realm.
+    String stateSignature = makeString(
```
I might be a bit confused but - if `generateRealmIdForFrame` should never be called with a null `documentID` if `isFrameExecutionReady` is always checked first. But `createRealmInfoForFrame` calls `generateRealmIdForFrame` without checking. Is this intentional ?

sergedeh replied:
Currently, the `createRealmInfoFrame` is only called within a `isFrameExecutionReady ` guarded block in `collectExecutionReadyFrameRealms`. I will add a precondition within createRealmInfoForFrame to prevent any accidental misuse.

AnujPanta1 commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,311 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+    return makeString("realm-"_s, browsingContext);
+}
+
+String BidiScriptAgent::originStringFromSecurityOriginData(const WebCore::SecurityOriginData& originData)
```
This could be static member

AnujPanta1 commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,311 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+    return newRealmId;
+}
+
+String BidiScriptAgent::generateRealmIdForBrowsingContext(const String& browsingContext)
```
Could be a static member

sergedeh replied:
Agreed. As it is pure and doesn't make use of any instance, I will update it to static. When/If the new version requires access to session/frame data, static will be dropped at that time.

AnujPanta1 commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,311 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+    // Per W3C BiDi spec §7.6.3.11, worklet realms should also be returned.
+
+    RefPtr session = m_session.get();
+    if (!session) {
```
I'd prefer macros like you used else where in the PR just so we have a consistent style

sergedeh replied:
Understood. I’ll align this with the existing early-return macro.

AnujPanta1 commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,311 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+    Ref<WebPageProxy> currentPage = pagesToProcess[0];
+    pagesToProcess.removeAt(0);
+
+    currentPage->getAllFrameTrees([this, pagesToProcess = WTFMove(pagesToProcess), optionalType = WTFMove(optionalType), contextHandleFilter = WTFMove(contextHandleFilter), accumulated = WTFMove(accumulated), callback = WTFMove(callback)](Vector<FrameTreeNodeData>&& frameTrees) mutable {
```
You capture `this` in an async callback. If `BidiScriptAgent` is destroyed before the callback fires,
you'll have a dangling pointer, maybe something like this 
```
  currentPage->getAllFrameTrees([weakThis = WeakPtr { *this }, ...](Vector<FrameTreeNodeData>&&
  frameTrees) mutable {
      RefPtr protectedThis = weakThis.get();
      if (!protectedThis)
          return;
      // ...
  });
```

sergedeh replied:
Agreed.

AnujPanta1 commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,311 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+
+        // Extract the URL from the cached state signature (format: "URL-doc-DOCID")
+        String cachedURL = cachedStateSignature;
+        size_t docPos = cachedStateSignature.find("-doc-"_s);
```
I have some questions about this
1. What if the URL itself contains -doc-? e.g., https://example.com/my-doc-page - seems fragile
2. Format knowledge seems to be split in two places, when you create stateSignature we have the `URL + "-doc-" + docID`, `.find("-doc-"_s)` assume a format when parsing

Do you think a structure might be more beneficial ? 

```
  struct FrameRealmCacheEntry {
      String url;
      String documentID;
      String realmId;
  };
  HashMap<WebCore::FrameIdentifier, FrameRealmCacheEntry> m_frameRealmCache;
```

```
  // caching:
  m_frameRealmCache.set(frameInfo.frameID, FrameRealmCacheEntry {
      frameInfo.request.url().string(),
      frameInfo.documentID ? frameInfo.documentID->toString() : String(),
      newRealmId
  });

  // reading:
  const auto& cached = it->value;
  String cachedURL = cached.url;
  String cachedDocID = cached.documentID; 

```

sergedeh replied:
1. Good point. Parsing the '-doc-' is indeed brittle.
2. Agreed, I will switch the cache entry to struct as suggested for a more robust solution.

AnujPanta1 commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,311 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+    }
+
+    // Recurse into subframes if requested
+    if (recurseSubframes) {
```
I'm curious about this conditional, from what I see we always call `collectExecutionReadyFrameRealms(frameTree, candidateRealms, contextHandleFilter, false);` with a false so this never triggers, is this for a future support if so maybe a FIXME for explanation ? Seems like a dead parameter

sergedeh replied:
It is actually for future support. I will add a FIXME to clarify this.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,290 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
 
+void BidiScriptAgent::getRealms(const String& optionalContext, std::optional<Inspector::Protocol::BidiScript::RealmType>&& optionalType, Inspector::CommandCallback<Ref<JSON::ArrayOf<Inspector::Protocol::BidiScript::RealmInfo>>>&& callback)
+{
+    // FIXME: Implement worker realm support (dedicated-worker, shared-worker, service-worker, worker)
```
I'd prefer you link to the draft spec section for the revelant endpoint once up at the top, it will be easier to follow subsequent section numbers. 

Also just a heads up that spec section numbers do occasionally change, so my preference is to omit them unless it's unclear or multiple sections are referenced.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,290 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+        else
+            ASYNC_FAIL_WITH_PREDEFINED_ERROR(FrameNotFound);
+
+        if (!resolvedPageForContext)
```
Nit: use the `_IF` variant of the macro to avoid needing the branch.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,290 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+    // Currently only window realms (main frames and iframes) are supported.
+    // Worker realm types require tracking worker global scopes and their owner sets.
+
+    // FIXME: Implement worklet realm support (paint-worklet, audio-worklet, worklet)
```
Please file bugzilla bugs for each of the FIXME next steps and add the links. It's OK if this results in one or two followups, it's just easier to track what's left if we create bugs along the way.

```
// FIXME: <webkit.org/b/123456> implement foobar
```

or on a separate line if the comment is a full line long.

```
// FIXME: something something something
// https://webkit.org/b/123456
```

 Then once those issues are fixed in a later PR, the FIXME with bugzilla link will get removed.

burg commented on metadata/contributors.json:
```diff
@@ -7550,6 +7550,18 @@
+         "serge.deh@gitstart.dev",
+         "sergedeh@gmail.com"
+      ],
+      "expertise" : "WebDriver",
```
Welcome! You can (and should!) make this part into a separate PR and move forward with it.

Later on, if this `script.getRealms` PR is landed but needs to be reverted, we wouldn't want your contact info to also get reverted along with it.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.h:
```diff
@@ -29,28 +29,63 @@
     // Inspector::BidiScriptBackendDispatcherHandler methods.
     void callFunction(const String& functionDeclaration, bool awaitPromise, Ref<JSON::Object>&& target, RefPtr<JSON::Array>&& optionalArguments, std::optional<Inspector::Protocol::BidiScript::ResultOwnership>&&, RefPtr<JSON::Object>&& optionalSerializationOptions, RefPtr<JSON::Object>&& optionalThis, std::optional<bool>&& optionalUserActivation, Inspector::CommandCallbackOf<Inspector::Protocol::B ...
     void evaluate(const String& expression, bool awaitPromise, Ref<JSON::Object>&& target, std::optional<Inspector::Protocol::BidiScript::ResultOwnership>&&, RefPtr<JSON::Object>&& optionalSerializationOptions,  std::optional<bool>&& optionalUserActivation, Inspector::CommandCallbackOf<Inspector::Protocol::BidiScript::EvaluateResultType, String, RefPtr<Inspector::Protocol::BidiScript::RemoteValue ...
+    void getRealms(const String& optionalContext, std::optional<Inspector::Protocol::BidiScript::RealmType>&& optionalType, Inspector::CommandCallback<Ref<JSON::ArrayOf<Inspector::Protocol::BidiScript::RealmInfo>>>&&) override;
```
For the first argument, I prefer we use the typedef `Inspector::Protocol::BidiBrowsingContext::BrowsingContext` and call it `optionalBrowsingContext` to distinguish from user contexts, which is a separate concept.

In the .cpp file it's OK to shorten this by adding

```
using BrowsingContext = Inspector::Protocol::BidiBrowsingContext::BrowsingContext;
```

sergedeh replied:
Thank you! This is definitely clearer.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.h:
```diff
@@ -29,28 +29,63 @@
+    void getRealms(const String& optionalContext, std::optional<Inspector::Protocol::BidiScript::RealmType>&& optionalType, Inspector::CommandCallback<Ref<JSON::ArrayOf<Inspector::Protocol::BidiScript::RealmInfo>>>&&) override;
 
 private:
+    void processRealmsForPagesAsync(Deque<Ref<WebPageProxy>>&& pagesToProcess, std::optional<Inspector::Protocol::BidiScript::RealmType>&& optionalType, std::optional<String>&& contextHandleFilter, Vector<RefPtr<Inspector::Protocol::BidiScript::RealmInfo>>&& accumulated, Inspector::CommandCallback<Ref<JSON::ArrayOf<Inspector::Protocol::BidiScript::RealmInfo>>>&&);
```
Nit: `optionalRealmType` is going to be so much easier to search for later.

burg commented on WebDriverTests/TestExpectations.json:
```diff
@@ -3811,20 +3811,34 @@
-    },
     "imported/w3c/webdriver/tests/bidi/script/get_realms/get_realms.py": {
-        "expected": { "all": { "status": ["SKIP"], "bug": "webkit.org/b/288061"}}
+        "subtests": {
```
You did a great job documenting the next steps here with links.

sergedeh replied:
Thanks! Appreciate the feedback.

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiScript.json:
```diff
@@ -67,6 +67,26 @@
+            "description": "The type of a JavaScript realm.",
+            "spec": "https://w3c.github.io/webdriver-bidi/#type-script-RealmType",
+            "type": "string",
+            "enum": [ "window", "dedicated-worker", "shared-worker", "service-worker", "worker", "paint-worklet", "audio-worklet", "worklet" ]
```
Please break the enum values onto one line each. See `Automation.json` for examples.

There's the question as to whether we should add worklet-related values if they aren't actually supported yet. Are these values actually used?

Please remove the ones that aren't getting used yet, or add a note in the description to explain the current state.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -91,7 +105,7 @@ void BidiScriptAgent::callFunction(const String& functionDeclaration, bool await
                     .release();
 
-                callback({ { Inspector::Protocol::BidiScript::EvaluateResultType::Exception, "placeholder_realm"_s, nullptr, WTFMove(exceptionDetails) } });
+                callback({ { Inspector::Protocol::BidiScript::EvaluateResultType::Exception, realmId, nullptr, WTFMove(exceptionDetails) } });
```
Nit: I prefer these be spelled `realmID`

burg replied:
I recommend a `using EvaluateResultType = ...` at the top to shorten subsequent uses of the enum.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,290 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+
+    // For main frames, return the page handle
+    if (!frameInfo.isMainFrame)
+        return emptyString(); // Only handling main frames in this PR
```
Replace this comment with FIXME above the `if`

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,290 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+    }
+
+    // Generate a new realm ID - the state has changed or this is a new frame
+    String contextHandle = contextHandleForFrame(frameInfo);
```
Since this code isn't directly protocol-exposed, I'd really prefer we use `std::optional<String>` rather than relying on the convention of `emptyString()` being nullopt.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,290 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+        if (cachedEntry.url == currentURL && cachedEntry.documentID == currentDocumentID)
+            return cachedEntry.realmId;
+
+        // If only the documentID changed but URL is the same, this is likely a timing issue
```
It's not quite clear what situation would lead to this. Did you investigate what the timing issue might be?

sergedeh replied:
During page reload (URL being the same), WebKit creates a new Document with a new documentID and a new Realm (JSGlobalObject) (see [DocumentWriter::begin](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/loader/DocumentWriter.cpp#L149-L158)). Since realm lifecycle events have not been implemented (https://bugs.webkit.org/show_bug.cgi?id=304062), UIProcess has no notification when WebProcess creates/destroys realms. The cache and counter manage the mapping between WebKit's documentID (which changes) and BiDi's realm ID (which we want stable). The lazy creation of realms in WebProcess, combined with the lack of lifecycle events, creates timing issues where documentID can change between calls while the URL remains constant. Conversely, documentID can even remain constant when a genuine navigation occurs, though this is an edge case better addressed by the realm lifecycle events implementation.
 Updated the comment to clarify this as a workaround and reference the bug.

burg replied:
Okay. That's a great explanation!

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -141,12 +155,290 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+    String currentURL = frameInfo.request.url().string();
+    std::optional<String> currentDocumentID = frameInfo.documentID ? std::optional<String>(frameInfo.documentID->toString()) : std::nullopt;
+
+    // Check if we have a cached realm ID for this frame
```
First two comments here are unnecessary.

burg reviewed (commented):
Overall this is looking good, just some questions and style suggestions.

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
```diff
@@ -30,11 +30,24 @@
+#include "WebFrameProxy.h"
 #include "WebPageProxy.h"
+#include "WebProcessPool.h"
+
```
Nit: no newline here

burg commented on Source/WebKit/UIProcess/Automation/BidiScriptAgent.h:
```diff
@@ -29,28 +29,63 @@
 namespace WebKit {
 
 class WebAutomationSession;
+class WebPageProxy;
```
Nit: alphabetize

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiScript.json:
```diff
@@ -67,6 +67,26 @@
+                { "name": "origin", "type": "string", "description": "The serialized origin of the realm." },
+                { "name": "type", "$ref": "BidiScript.RealmType", "description": "The type of the realm." },
+                { "name": "context", "$ref": "BidiBrowsingContext.BrowsingContext", "optional": true, "description": "Browsing context of the realm. Not present for worklet/worker realms." },
+                { "name": "sandbox", "type": "string", "optional": true, "description": "Sandbox name. Only present for sandbox realms." }
```
Since we don't support this parameter and it's not really standardized, I'd like to drop "sandbox" parameter from this PR. It's optional, so we can add it back later if needed.

burg reviewed (approved):
I think we are ready to move forward. Yay!

sergedeh committed:
Implement script.getRealms for main window contexts
https://bugs.webkit.org/show_bug.cgi?id=303926

Reviewed by BJ Burg.

This change implements the WebDriver BiDi script.getRealms command for main window browsing contexts, enabling tests to query JavaScript realms in the page.

Added RealmInfo, RealmType types to BidiScript.json protocol, implemented getRealms() in BidiScriptAgent for main window contexts with realm ID generation and navigation detection. Added basic realm ID support to evaluate() and callFunction().

Note: Realm ID consistency between evaluate() and getRealms() is not fully implemented - this will be addressed in a follow-up change.

* Source/WebKit/UIProcess/Automation/BidiScriptAgent.h:
* Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp:
(WebKit::BidiScriptAgent::callFunction):
(WebKit::BidiScriptAgent::evaluate):
(WebKit::BidiScriptAgent::getRealms):
(WebKit::BidiScriptAgent::createRealmInfoForFrame):
(WebKit::BidiScriptAgent::generateRealmIdForFrame):
(WebKit::BidiScriptAgent::generateRealmIdForBrowsingContext):
(WebKit::BidiScriptAgent::processRealmsForPagesAsync):
(WebKit::BidiScriptAgent::contextHandleForFrame):
(WebKit::BidiScriptAgent::collectExecutionReadyFrameRealms):
* Source/WebKit/UIProcess/Automation/protocol/BidiScript.json:
* WebDriverTests/TestExpectations.json:

Canonical link: https://commits.webkit.org/304800@main
